src/annotation/inject_base.ts

```diff
@@ -0,0 +1,19 @@
+import { UNDEFINED_INJECT_ANNOTATION } from "../constants/error_msgs";
+import { Metadata } from "../planning/metadata";
+import { createTaggedDecorator } from "./decorator_utils";
+import { ServiceIdentifierOrFunc } from "./lazy_service_identifier";
+
+export function injectBase(metadataKey:string){
+  return (serviceIdentifier: ServiceIdentifierOrFunc) => {
+    return (target: Object,
```

Comment:
I think `Record<string | symbol, unknown> | interfaces.Newable<unknown>` is a better type for `target`

Reply:
Not acceptable by the compiler

> is not assignable to type 'Record<string | symbol, unknown>'. Index signature is missing in type

I think that I have provided a solution

Reply:
```
interface ConstructorFunction {
    new(...args:any[]):unknown,
    readonly prototype:unknown
}
interface Prototype {
    constructor: Function;
}
export type DecoratorTarget = Prototype | ConstructorFunction;
```

Reply:
It's indeed a valid solution. We want to introduce some linter rules in the future and that's the reason to [ban some types](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-types.md). If you want ÇI can create a PR to your branch with equivalent types which respect this rule :)

Reply:
> Can you elaborate on CI ?

I assume you mean I. So go ahead !

Reply:
Yeah, that was a typo

Reply:
I spent a couple of hours wondering how to improve typings because my original idea didn't work. At the end `object` is good enough imo, it covers all the cases and respects the `Reflect` typings. Sorry for the waste of time

Reply:
But in the linked "ban some types" the object type is problematic too.

Also, although object ( and Object ) is allowed for Reflect typings what we are adding metadata to is the constructor function and as such the annotationTarget should be Function in `_tagParameterOrProperty`.

```
class MetadataReader implements interfaces.MetadataReader {

    public getConstructorMetadata(constructorFunc: Function): interfaces.ConstructorMetadata {

        // TypeScript compiler generated annotations
        const compilerGeneratedMetadata = Reflect.getMetadata(METADATA_KEY.PARAM_TYPES, constructorFunc);

        // User generated constructor annotations
        const userGeneratedMetadata = Reflect.getMetadata(METADATA_KEY.TAGGED, constructorFunc);

        return {
            compilerGeneratedMetadata,
            userGeneratedMetadata: userGeneratedMetadata || {}
        };

    }

    public getPropertiesMetadata(constructorFunc: Function): interfaces.MetadataMap {
        // User generated properties annotations
        const userGeneratedMetadata =  Reflect.getMetadata(METADATA_KEY.TAGGED_PROP, constructorFunc) || [];
        return userGeneratedMetadata;
    }

}

```
 
We then need to ensure that tagParameter passes a function.  I suggest that we change tagParameter to below.
We do not need to check the property to known that the decorator is on a method parameter as the annotation target tells us.
```
function tagParameter(
    annotationTarget: object,
    _: string | symbol | undefined,
    parameterIndex: number,
    metadata: interfaces.MetadataOrMetadataArray
) {
    if ( typeof annotationTarget === "function") {
        _tagParameterOrProperty(METADATA_KEY.TAGGED, annotationTarget, parameterIndex.toString(), metadata);
    }else{
        throw new Error(ERROR_MSGS.INVALID_DECORATOR_OPERATION);
    }
}
```

Reply:
The name is required for static method parameters. Take that back.

Reply:
we can also just add [this ](https://palantir.github.io/tslint/rules/ban-types/)into out `tslint` setup right now and then just try out what works best, since `ban-types` are somehow related to both linters.

What was the problem with `Record<string, unknown` anyway ?

Reply:
> What was the problem with Record<string, unknown anyway ?

Because it is incorrect.  
One, the target can be a function
in injectBase
```
const className = typeof target === "function" ? target.name : target.constructor.name;
```
Secondly when applied to a property typescript will complain

Reply:
https://github.com/palantir/tslint/blob/285fc1db18d1fd24680d6a2282c6445abf1566ee/src/configs/recommended.ts#L23

Reply:
@PodaruDragos 
tell eslint to ignore the object and the Function types where they are used.  The errors is there for lazy typing and errors that arise because of that.  
In this instance the type signatures are ok and definitely better than unknown and of couse any.

Reply:
If the class being decorated had
`[key:string]:unknown`
Then `Record<string | symbol, unknown> | interfaces.Newable<unknown>` will not error